src/Microsoft.OData.Core/ODataMessageReader.cs

```diff
@@ -887,29 +888,29 @@ internal ODataFormat GetFormat()
             return this.format;
         }
 
-        private static IServiceProvider GetContainer<T>(T message)
+        private static IServiceProvider GetServiceProvider<T>(T message)
             where T : class
         {
-            var containerProvider = message as IContainerProvider;
-            return containerProvider == null ? null : containerProvider.Container;
+            var serviceCollectionProvider = message as IServiceCollectionProvider;
+            return serviceCollectionProvider?.ServiceProvider;
         }
 
-        private static IEdmModel GetModel(IServiceProvider container)
+        private static IEdmModel GetModel(IServiceProvider serviceProvider)
         {
-            return container == null ? EdmCoreModel.Instance : container.GetRequiredService<IEdmModel>();
+            return serviceProvider == null ? EdmCoreModel.Instance : serviceProvider.GetRequiredService<IEdmModel>();
         }
 
         private ODataMessageInfo GetOrCreateMessageInfo(Stream messageStream, bool isAsync)
         {
             if (this.messageInfo == null)
             {
-                if (this.container == null)
+                if (this.serviceProvider == null)
                 {
                     this.messageInfo = new ODataMessageInfo();
                 }
                 else
                 {
-                    this.messageInfo = this.container.GetRequiredService<ODataMessageInfo>();
+                    this.messageInfo = this.serviceProvider.GetRequiredService<ODataMessageInfo>();

```

Comment:
Given that `messageInfo` is modified in this method, actually all of it's properties are modified, there seems to be reason to get this from the service provider.

If the caller provided a custom `ODataMessageInfo`, then we would overwrite all the properties they set, making it pointless (and arguably an anti-pattern).

I think we should remove this class from the service container and consider making it private.

I think customers who have passed a service provider without creating a scope per request have ran into issues because threads from different requests ended up writing to the same message stream. Of course, this service was designed to be scoped, so that's technically a customer error. But my argument is that it shouldn't be in the DI container in the first place.

Reply:
To avoid derailing this PR. I've created a separate issue to track and discuss this: https://github.com/OData/odata.net/issues/2908